pkg/ingester/ingester_v2.go

```diff
@@ -110,6 +112,37 @@ func (u *userTSDB) PostDeletion(metrics ...labels.Labels) {
 	}
 }
 
+// blocksToDelete filters the input blocks and returns the blocks which are safe to be deleted from the ingester.
+func (u *userTSDB) blocksToDelete(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
+	if u.DB == nil {
+		return map[ulid.ULID]struct{}{}
+	}
+	deletable := tsdb.DefaultBlocksToDelete(u.DB)(blocks)
+	if u.shipper == nil {
+		return deletable
+	}
+
+	shipperMeta, err := shipper.ReadMetaFile(u.Dir())
+	if err != nil {
+		// If there is any issue with the shipper, we should be conservative and not delete anything.
+		level.Error(util.Logger).Log("msg", "failed to read shipper meta during deletion of blocks", "user", u.userID, "err", err)
+		return map[ulid.ULID]struct{}{}
+	}
+
+Outer:
+	for id := range deletable {
+		for _, shippedID := range shipperMeta.Uploaded {
+			if bytes.Equal(shippedID[:], id[:]) {
+				continue Outer
+			}
+		}
+		// Not shipped yet.
+		delete(deletable, id)
+	}
+
+	return deletable
```

Comment:
Can be simplified to:

```suggestion
	result := map[ulid.ULID]struct{}{}
	for _, shippedID := range shipperMeta.Uploaded {
		if _, ok := deletable[shippedID]; ok {
			result[shippedID] = struct{}{}
		}
	}
	return result
```